substrate/frame/support/procedural/src/pallet/expand/tasks.rs

```diff
@@ -0,0 +1,257 @@
+					#scrate::EqNoBound,
+					#scrate::PartialEqNoBound,
+					#scrate::pallet_prelude::Encode,
+					#scrate::pallet_prelude::Decode,
+					#scrate::pallet_prelude::TypeInfo,
+				)]
+				#[codec(encode_bound())]
+				#[codec(decode_bound())]
+				#[scale_info(skip_type_params(#type_use_generics))]
+				#vis enum #ident #generics {
+					#variants
+					#[doc(hidden)]
+					#[codec(skip)]
+					__Ignore(core::marker::PhantomData<T>, #scrate::Never),
+				}
+
+				impl<T: Config> core::fmt::Debug for #ident<#type_use_generics> {
+					fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+						f.debug_struct(stringify!(#ident)).field("value", self).finish()
+					}
+				}
+			});
+		} else {
+			// `item_enum` is a manually specified enum (no attribute)
+			tokens.extend(item_enum.to_token_stream());
+		}
+	}
+}
+
+/// Represents an already-expanded [`TasksDef`].
+#[derive(Parse)]
+pub struct ExpandedTasksDef {
+	pub task_item_impl: ItemImpl,
+	pub task_trait_impl: ItemImpl,
+}
+
+impl ToTokens for TasksDef {
+	fn to_tokens(&self, tokens: &mut TokenStream2) {
+		let scrate = &self.scrate;
+		let enum_ident = syn::Ident::new("Task", self.enum_ident.span());
+		let enum_arguments = &self.enum_arguments;
+		let enum_use = quote!(#enum_ident #enum_arguments);
+
+		let task_fn_idents = self
+			.tasks
+			.iter()
+			.map(|task| format_ident!("{}", &task.item.sig.ident.to_string().to_class_case()))
+			.collect::<Vec<_>>();
+		let task_indices = self.tasks.iter().map(|task| &task.index_attr.meta.index);
+		let task_conditions = self.tasks.iter().map(|task| &task.condition_attr.meta.expr);
+		let task_weights = self.tasks.iter().map(|task| &task.weight_attr.meta.expr);
+		let task_iters = self.tasks.iter().map(|task| &task.list_attr.meta.expr);
+
+		let task_fn_impls = self.tasks.iter().map(|task| {
+			let mut task_fn_impl = task.item.clone();
+			task_fn_impl.attrs = vec![];
+			task_fn_impl
+		});
+
+		let task_fn_names = self.tasks.iter().map(|task| &task.item.sig.ident);
+		let task_arg_names = self.tasks.iter().map(|task| &task.arg_names).collect::<Vec<_>>();
+
+		let sp_std = quote!(#scrate::__private::sp_std);
+		let impl_generics = &self.item_impl.generics;
+		tokens.extend(quote! {
+			impl #impl_generics #enum_use
+			{
+				#(#task_fn_impls)*
+			}
+
+			impl #impl_generics #scrate::traits::Task for #enum_use
+			{
+				type Enumeration = #sp_std::vec::IntoIter<#enum_use>;
+
+				fn iter() -> Self::Enumeration {
+					let mut all_tasks = #sp_std::vec![];
+					#(all_tasks
+						.extend(#task_iters.map(|(#(#task_arg_names),*)| #enum_ident::#task_fn_idents { #(#task_arg_names: #task_arg_names.clone()),* })
+						.collect::<#sp_std::vec::Vec<_>>());
+					)*
+					all_tasks.into_iter()
+				}
+
+				fn task_index(&self) -> u32 {
+					match self.clone() {
+						#(#enum_ident::#task_fn_idents { .. } => #task_indices,)*
+						Task::__Ignore(_, _) => unreachable!(),
+					}
+				}
+
+				fn is_valid(&self) -> bool {
```

Comment:
`valid` means that the task is ready to run? *Valid* sounds more like it is well defined or something.  
Or maybe `can_run` since we use the word `run` already.

Reply:
Tasks are a bit like transactions in that they're `dispatchable` and can exist as an extrinsic. The main differences are that:
- Tasks are free to run;
- All valid tasks can be enumerated;
- It can always be cheaply determined whether a task will execute successfully.

We already use the term "valid" for transactions which are allowed to be included on-chain (e.g. `TransactionValidity`, `ValidTransation`). I don't think it's a stretch to use it to mean the same thing for tasks too.